packages/flutter/lib/src/widgets/undo_history.dart

```diff
@@ -0,0 +1,466 @@
+        _updateState();
+      },
+    );
+    _push();
+    widget.value.addListener(_push);
+    _handleFocus();
+    widget.focusNode.addListener(_handleFocus);
+    _effectiveController.onUndo.addListener(undo);
+    _effectiveController.onRedo.addListener(redo);
+  }
+
+  @override
+  void didUpdateWidget(UndoHistory<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.value != oldWidget.value) {
+      _stack.clear();
+      oldWidget.value.removeListener(_push);
+      widget.value.addListener(_push);
+    }
+    if (widget.focusNode != oldWidget.focusNode) {
+      oldWidget.focusNode.removeListener(_handleFocus);
+      widget.focusNode.addListener(_handleFocus);
+    }
+    if (widget.controller != oldWidget.controller) {
+      _effectiveController.onUndo.removeListener(undo);
+      _effectiveController.onRedo.removeListener(redo);
+      _controller?.dispose();
+      _controller = null;
+      _effectiveController.onUndo.addListener(undo);
+      _effectiveController.onRedo.addListener(redo);
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.value.removeListener(_push);
+    widget.focusNode.removeListener(_handleFocus);
+    _effectiveController.onUndo.removeListener(undo);
+    _effectiveController.onRedo.removeListener(redo);
+    _controller?.dispose();
+    _throttleTimer?.cancel();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Actions(
+      actions: <Type, Action<Intent>>{
+        UndoTextIntent: Action<UndoTextIntent>.overridable(context: context, defaultAction: CallbackAction<UndoTextIntent>(onInvoke: _undoFromIntent)),
+        RedoTextIntent: Action<RedoTextIntent>.overridable(context: context, defaultAction: CallbackAction<RedoTextIntent>(onInvoke: _redoFromIntent)),
+      },
+      child: widget.child,
+    );
+  }
+}
+
+/// Represents whether the current undo stack can undo or redo.
+@immutable
+class UndoHistoryValue {
+  /// Creates a value for whether the current undo stack can undo or redo.
+  ///
+  /// The [canUndo] and [canRedo] arguments must have a value, but default to
+  /// false.
+  const UndoHistoryValue({this.canUndo = false, this.canRedo = false});
+
+  /// A value corresponding to an undo stack that can neither undo nor redo.
+  static const UndoHistoryValue empty = UndoHistoryValue();
+
+  /// Whether the current undo stack can perform an undo operation.
+  final bool canUndo;
+
+  /// Whether the current undo stack can perform a redo operation.
+  final bool canRedo;
+
+  @override
+  String toString() => '${objectRuntimeType(this, 'UndoHistoryValue')}(canUndo: $canUndo, canRedo: $canRedo)';
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) {
+      return true;
+    }
+    return other is UndoHistoryValue && other.canUndo == canUndo && other.canRedo == canRedo;
+  }
+
+  @override
+  int get hashCode => Object.hash(
+        canUndo.hashCode,
+        canRedo.hashCode,
+      );
```

Comment:
Nit: I could be off but I think this way is most aligned with Flutter's code style:

```dart
  int get hashCode => Object.hash(
    canUndo.hashCode,
    canRedo.hashCode,
  );
```